File: src/main/java/com/trip_service/trip_service/security/middlewares/AuthenticationMiddleware.java
package com.trip_service.trip_service.security.middlewares;

import com.fasterxml.jackson.databind.ObjectMapper;
import jakarta.servlet.ServletException;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import org.springframework.security.core.AuthenticationException;
import org.springframework.security.web.AuthenticationEntryPoint;
import org.springframework.stereotype.Component;

import java.io.IOException;
import java.util.HashMap;
import java.util.Map;

@Component
public class AuthenticationMiddleware implements AuthenticationEntryPoint {

    @Override
    public void commence(HttpServletRequest request, HttpServletResponse response, AuthenticationException authException) throws IOException, ServletException {
        response.setStatus(HttpServletResponse.SC_UNAUTHORIZED); // 401 Unauthorized
        response.setContentType("application/json");

        // Create JSON response body
        Map<String, Object> errorResponse = new HashMap<>();
        errorResponse.put("status", 401);
        errorResponse.put("error", "Unauthorized");
        errorResponse.put("message", authException.getMessage()); // Shows reason (Bad Credentials, Token Expired, etc.)
        errorResponse.put("path", request.getRequestURI());

        // Convert Map to JSON
        ObjectMapper objectMapper = new ObjectMapper();
        String jsonResponse = objectMapper.writeValueAsString(errorResponse);

        // Send JSON response
        response.getWriter().write(jsonResponse);
    }
}

File: src/main/kotlin/com/trip_service/trip_service/security/middlewares/AuthenticationMiddleware.kt
package com.trip_service.trip_service.security.middlewares

import com.fasterxml.jackson.databind.ObjectMapper
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.springframework.security.core.AuthenticationException
import org.springframework.security.web.AuthenticationEntryPoint
import org.springframework.stereotype.Component

@Component
class AuthenticationMiddleware : AuthenticationEntryPoint {

    override fun commence(
        request: HttpServletRequest?,
        response: HttpServletResponse?,
        authException: AuthenticationException?
    ) {

        response?.status = HttpServletResponse.SC_UNAUTHORIZED;
        response?.contentType = "application/json";

        // Create JSON Response Body
        val errorResponse = mutableMapOf<String, Any>();

        errorResponse["status"] =  401;
        errorResponse["error"] =  "Unauthorized";
        errorResponse["message"] = authException?.message ?: "";
        errorResponse["path"] =  request?.requestURI ?: "";

        // Convert Map to JSON
        val objectMapper: ObjectMapper = ObjectMapper();
        val jsonResponse: String = objectMapper.writeValueAsString(errorResponse);

        // Send Json Response
        response?.writer?.write(jsonResponse);

    }

}